src/core/unif.ml

```diff
@@ -32,13 +56,12 @@ let add_constr : problem -> constr -> unit = fun p c ->
    as well as the constraint [(c,a,b)] where [a] is the type of [t] and [b]
    the type of [u] if they can be infered. *)
 let add_unif_rule_constr : problem -> constr -> unit = fun p (c,t,u) ->
-  add_constr p (c,t,u);
   match Infer.infer_noexn p c t with
-  | None -> ignore (Infer.infer_noexn p c u)
-  | Some a ->
+  | None -> add_constr p (c, t, u)
+  | Some (t, a) ->
       match Infer.infer_noexn p c u with
-      | Some b when not (Eval.pure_eq_modulo c a b) -> add_constr p (c,a,b)
-      | _ -> ()
+      | None -> add_constr p (c, t, u)
+      | Some (u, b) -> add_constr p (c, t, u); add_constr p (c,a,b)
```

Comment:
Why not keeping the test (Eval.pure_eq_modulo c a b)?

Reply:
Because even if `a` and `b` are equal modulo, term `u` may have been  refined. But what we may do is
```ocaml
| Some (u, b) -> add_constr p (c, t, u); if not (Eval.pure_eq_modulo c a b) then add_constr p (c, a, b)
```

Reply:
Ok.